---
title: "Readme"
author: "Nidia Barco Armengol"
output: github_document
---

# Introduction

The Deconvolution package contains different functions, specific to generate absolute or relative deconvolution plots, with different methods, which also return data frames with the result of the deconvolution, or a list with data frames in the case of the deconv function, these df can then be used with another function of the package, the Box_Deconv function to generate a boxplot that compares two conditions in the different cell types, and does so from the df resulting from the deconvolution. 

# Methods

| Deconvolution method                                           | Description                                                                                                                                                                                                                                                                                                                                                                                                                                         |
|----------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CIBERSORT                                                      | Versatile computational method for quantifying cell fractions from bulk tissue gene expression profiles (GEPs). Combining support vector regression with prior knowledge of expression profiles from purified leukocyte subsets,this methode can accurately estimate the immune composition of a tumor biopsy.                                                                                                                                      |
| EPIC (Estimating the Proportions of Immune and Cancer cells)   | includes RNA-seq-based gene expression reference profiles from immune cells and other nonmalignant cell types found in tumors, can additionally manage user-defined gene expression reference profiles, include the ability to account for an uncharacterized cell type.                                                                                                                                                                            |
| FARDEEP (Fast And Robust DEconvolution of Expression Profiles) | This methode enumerate immune cell subsets from whole tumor tissue samples. To reduce noise in the tumor gene expression datasets, utilizes an adaptive least trimmed square to automatically detect and remove outliers before estimating the cell compositions. We show that FARDEEP is less susceptible to outliers and returns a better estimation of coefficients than the existing methods with both numerical simulations and real datasets. |

# Signature matrix

| Signature matrix                | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
|---------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LM22                            | Is a signature matrix file consisting of 547 genes that accurately distinguish 22 mature human hematopoietic populations isolated from peripheral blood or in vitro culture conditions, including seven T cell types, na√Øve and memory B cells, plasma cells, NK cells, and myeloid subsets.                                                                                                                                                                                                              |
| Monaco-ABIS_Immune11_Microarray | This matrix was constructed from a set of predictor variables (cell types), normalized by mRNA abundance with scaling factors, and filtered by extracting genes with very high or very low expression and also genes not expressed with the STAR method, including eleven cell tipes, T.Naive, T.Memory, B.Naive, B.Memory, Plasmablasts, NK, pDCs, Neutrophils.LD, Basophils.LD, mDCs, Monocytes.                                                                                                        |
| Wang2020_signature.matrix       | To construct the signature matrix of brain data, adult scRNA-seq data from Darmanis et al. (2015) were analyzed by grouping characterized human brain cells into seven cell types [astrocytes, endothelia, microglia, oligodendrocytes (Oligo), oligodendrocyte precursor cells (OPCs), inhibitory and excitatory neurons] and selecting the top 50 marker genes for each cell type using SC3 (Kiselev et al., 2017). Finally, the expression of each marker gene in cells of the same type was averaged. |

# Functions

| Function         | Description                                                                                                                                                         | Return                                    |
|------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------|
| CIBERSORT_abs_F  | Function that generates the absolute deconvolution plot with the CIBERSORT method, with the fractional plot, and returns a df resulting from the deconvolution.     | Deconvolution Graphs and Deconvolution df |
| CIBERSORT_abs_NF | Function that generates the absolute deconvolution plot with the CIBERSORT method, with the non-fractional plot, and returns a df resulting from the deconvolution. | Deconvolution Graph and df                |
| CIBERSORT_rel_F  | Function that generates the relative deconvolution plot with the CIBERSORT method, with the fractional plot, and returns a df resulting from the deconvolution.     | Deconvolution Graphs and Deconvolution df |
| CIBERSORT_rel_NF | Function that generates the relative deconvolution plot with the CIBERSORT method, with the non-fractional plot, and returns a df resulting from the deconvolution. | Deconvolution Graph and df                |
| EPIC_rel_F       | Function that generates the relative deconvolution plot with the EPIC method, with the fractional plot, and returns a df resulting from the deconvolution.          | Deconvolution Graphs and Deconvolution df |
| EPIC_rel_NF      | Function that generates the relative deconvolution plot with the EPIC method, with the non-fractional plot, and returns a df resulting from the deconvolution.      | Deconvolution Graph and df                |
| FARDEEP_abs_F    | Function that generates the absolute deconvolution plot with the FARDEEP method, with the fractional plot, and returns a df resulting from the deconvolution.       | Deconvolution Graphs and Deconvolution df |
| FARDEEP_abs_NF   | Function that generates the absolute deconvolution plot with the FARDEEP method, with the non-fractional plot, and returns a df resulting from the deconvolution.   | Deconvolution Graph and df                |
| FARDEEP_rel_F    | Function that generates the relative deconvolution plot with the FARDEEP method, with the fractional plot, and returns a df resulting from the deconvolution.       | Deconvolution Graphs and Deconvolution df |
| FARDEEP_rel_NF   | Function that generates the relative deconvolution plot with the FARDEEP method, with the non-fractional plot, and returns a df resulting from the deconvolution.   | Deconvolution Graph and df                |
| Deconv           | Function that generates massive deconvolution plots with the CIBERSORT, EPIC and FARDEEP methods, and returns a list of df's resulting from the deconvolutions.     | Deconvolution Graphs and df's list        |
| Box_Deconv       | Function that generates a boxplot comparing conditions for each cell type from a df and a condition vector, and in the case of more than two conditions, it also returns a df with P.Value information.                                                     | Deconvolution Boxplot                     |

# Installation

```R
library(devtools)
devtools::install_github("margenomics/Deconvolution")
library(Deconvolution)
```

#  Usage

```R
# CIBERSORT_abs_F
  C_abs <- CIBERSORT_abs_F("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", f, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name, "MONACO")

# CIBERSORT_abs_NF
  C_abs <- CIBERSORT_abs_NF("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# CIBERSORT_rel_F
  C_rel <- CIBERSORT_rel_F("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", f, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# CIBERSORT_rel_NF
  C_rel <- CIBERSORT_rel_NF("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# EPIC_rel_F
  E_rel <- EPIC_rel_F("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", f, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# EPIC_rel_NF
  E_rel <- EPIC_rel_NF("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# FARDEEP_abs_F
  F_abs <- FARDEEP_abs_F("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", f, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# FARDEEP_abs_NF
  F_abs <- FARDEEP_abs_NF("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# FARDEEP_rel_F
  F_rel <- FARDEEP_rel_F("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", f, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# FARDEEP_rel_NF
  F_rel <- FARDEEP_rel_NF("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# Deconv
  L_rel <- deconv("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", "ALL", "rel", fraccionate.samples=TRUE, fractions= f, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")

# Box_deconv
  P.value_df <- Box_Deconv(data= C_abs, cond= fractions, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", f_name= "Box_CIBERSORT_abs")
```

# Examples

```R
matrix <- "Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt"
sig.matrix <- "Monaco-ABIS_Immune11_Microarray.txt"
fractions <- c(rep("A", each=18), rep("B", each=43), rep("C", each=36))
results_dir <- "/bicoh/nidia/Deconv/Brain_MMartin/test"
```

## CIBERSORT_abs_NF

### Itroduce:

```R
C_abs <- CIBERSORT_abs_NF("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test")
```

### Return:

![Image text](/bicoh/nidia/Deconvolution/Images/CIBERSORT.abs_Wang2020_signature.matrix__plot.png)
```{r, include=FALSE}
source("CIBERSORT.R")
matrix <- "Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt"
sig.matrix <- "Wang2020_signature.matrix.txt"
results_dir <- "/bicoh/nidia/Deconv/Brain_MMartin/test"
res_ciber = CIBERSORT(sig_matrix=sig.matrix,
                        mixture_file = matrix,
                        QN = F,perm=100,absolute = T,abs_method='no.sumto1')
  res_ciber = as.data.frame(t(res_ciber))
  res_ciber$cell_type=rownames(res_ciber)
  C_abs <- res_ciber[1:(length(rownames(res_ciber))-4),]
```
```{r}
# Show the first 6 columns of the deconvolution df.
knitr::kable(C_abs[,1:6 ])
```

## CIBERSORT_rel_F

### Itroduce:

```R
C_rel <- CIBERSORT_rel_F("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", f, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test", height= 8, name= "MONACO")
```

### Return:

![Image text](/bicoh/nidia/Deconvolution/Images/C_rel_fractions.png)

```{r, include=FALSE}
source("CIBERSORT.R")
matrix <- "Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt"
sig.matrix <- "Wang2020_signature.matrix.txt"
results_dir <- "/bicoh/nidia/Deconv/Brain_MMartin/test"
res_ciber = CIBERSORT(sig_matrix=sig.matrix,
                        mixture_file = matrix,
                        QN = F,perm=100)
  res_ciber = as.data.frame(t(res_ciber))
  res_ciber$cell_type=rownames(res_ciber)
  C_rel <- res_ciber[1:(length(rownames(res_ciber))-4),]
```
```{r}
# Show the first 6 columns of the deconvolution df.
knitr::kable(C_rel[,1:6 ])
```

## Deconv

### Introduce

```R
L_rel <- deconv("Counts.HUMAN_nodup_Sal_Veh_IPs_INPUTs.txt", "Monaco-ABIS_Immune11_Microarray.txt", "ALL", "rel", fraccionate.samples=FALSE, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test")
```
### Return

![Image text](/bicoh/nidia/Deconvolution/Images/Ex_Deconv.png)

```{r, include=FALSE}
require(EPIC)
require(usethis)
require(devtools)
require(gplots)
require(data.table)
require(ggplot2)
require(dplyr)
require(tidyr)
require(immunedeconv)
require(RColorBrewer)
require(FARDEEP)
cpm=read.table(matrix,sep = "\t",header = T)

    cpm <- cpm %>%
      group_by(Gene.symbol) %>%
      summarise_all(mean)
    names <- cpm$Gene.symbol
    cpm=as.matrix(cpm[,-1])
    rownames(cpm)=names

  # Llegeix la matriu signature en txt i posa els rownames amb els noms de la columna de gens, 3 possibles noms
  sig.mtrx=read.table(sig.matrix,sep = "\t",header = T)
  
    sig.mtrx <- sig.mtrx %>%
      group_by(Gene.symbol) %>%
      summarise_all(mean)
    names <- sig.mtrx$Gene.symbol
    sig.mtrx=as.matrix(sig.mtrx[,-1])
    rownames(sig.mtrx)=names

  # Creaci√≥ de la llista de gens i la matriu signature i generaci√≥ de la matriu de deconvoluci√≥ i conversi√≥ d'aquesta a df
  C_EPIC <- list()
  C_EPIC[["sigGenes"]] <- rownames(sig.mtrx)
  C_EPIC[["refProfiles"]] <- as.matrix(sig.mtrx)

  RESULTS <- t(EPIC::EPIC(bulk=cpm, reference=C_EPIC, withOtherCells=TRUE, scaleExprs=FALSE)$cellFractions) #scaleExprs=TRUE by default: only keep genes in common between matrices

  RESULTS <- as.data.frame(RESULTS)
  RESULTS$cell_type=rownames(RESULTS)
  E_rel <- RESULTS
  
  # Construcci√≥ de la matriu de deconvoluci√≥ relativa i conversi√≥ d'aquesta en un dataframe
  RESULTS = t(FARDEEP::fardeep(sig.mtrx, cpm, nn = TRUE, intercept = TRUE, permn = 10, QN = FALSE)$relative.beta)
  #RESULTS = t(FARDEEP::fardeep(sig.mtrx, cpm, nn = TRUE, intercept = TRUE, permn = 10, QN = FALSE)$relative.beta)

  RESULTS <- as.data.frame(RESULTS)
  RESULTS$cell_type=rownames(RESULTS)
  F_rel <- RESULTS
  
  l_df <- list(C_rel, E_rel, F_rel)
```
```{r}
# l_df <- list(C_rel, E_rel, F_rel)
knitr::kable(l_df)
```

## Box_Deconv

### Introduce

```R
P.value_df <- Box_Deconv(data= C_abs, cond= fractions, results_dir = "/bicoh/nidia/Deconv/Brain_MMartin/test")
```
### Return

#### With 2 conditions

![Image text](/bicoh/nidia/Deconvolution/Images/2_cond_ex.png)

#### With more than 2 conditions

![Image text](/bicoh/nidia/Deconvolution/Images/3_cond_ex.png)
```{r, include=FALSE}
require(ggplot2)
require(ggpubr)
require(gplots)
require(data.table)
require(ggsignif)
require(tidyverse)
fractions <- c(rep("A", each=18), rep("B", each=43), rep("C", each=36))
data <- C_abs
cond <- fractions
data$cell_type <- NULL
data <- t(data)
test_m <- melt(as.data.table(data), variable.factor = FALSE)
condition <- c(rep(cond, times=length(colnames(data))))
df_m <- cbind(test_m, condition)

anno_df = compare_means(value ~ condition, group.by = "variable", data = df_m, method = "t.test")
```
```{r}
knitr::kable(anno_df)
```

# References

Hao, Y., Yan, M., Heath, B. R., Lei, Y. L., & Xie, Y. (2019). Fast and robust deconvolution of tumor infiltrating lymphocyte from expression profiles using least trimmed squares. PLoS computational biology, 15(5), e1006976. https://doi.org/10.1371/journal.pcbi.1006976

Chen, B., Khodadoust, M. S., Liu, C. L., Newman, A. M., & Alizadeh, A. A. (2018). Profiling Tumor Infiltrating Immune Cells with CIBERSORT. Methods in molecular biology (Clifton, N.J.), 1711, 243‚Äì259. https://doi.org/10.1007/978-1-4939-7493-1_12

Racle, J., & Gfeller, D. (2020). EPIC: A Tool to Estimate the Proportions of Different Cell Types from Bulk Gene Expression Data. Methods in molecular biology (Clifton, N.J.), 2120, 233‚Äì248. https://doi.org/10.1007/978-1-0716-0327-7_17

Avila Cobos, F., Alquicira-Hernandez, J., Powell, J.E. et al. Benchmarking of cell type deconvolution pipelines for transcriptomics data. Nat Commun 11, 5650 (2020). https://doi.org/10.1038/s41467-020-19015-1

Gregor Sturm, Francesca Finotello, Florent Petitprez, Jitao David Zhang, Jan Baumbach, Wolf H Fridman, Markus List, Tatsiana Aneichyk, Comprehensive evaluation of transcriptome-based cell-type quantification methods for immuno-oncology, Bioinformatics, Volume 35, Issue 14, July 2019, Pages i436‚Äìi445, https://doi.org/10.1093/bioinformatics/btz363

Francisco Avila Cobos, Jo Vandesompele, Pieter Mestdagh, Katleen De Preter, Computational deconvolution of transcriptomics data from mixed cell populations, Bioinformatics, Volume 34, Issue 11, 01 June 2018, Pages 1969‚Äì1979, https://doi.org/10.1093/bioinformatics/bty019

Finotello, F., Trajanoski, Z. Quantifying tumor-infiltrating immune cells from transcriptomics data. Cancer Immunol Immunother 67, 1031‚Äì1040 (2018). https://doi.org/10.1007/s00262-018-2150-z

Monaco, G., Lee, B., Xu, W., Mustafah, S., Hwang, Y. Y., Carr√©, C., Burdin, N., Visan, L., Ceccarelli, M., Poidinger, M., Zippelius, A., Pedro de Magalh√£es, J., & Larbi, A. (2019). RNA-Seq Signatures Normalized by mRNA Abundance Allow Absolute Deconvolution of Human Immune Cell Types. Cell reports, 26(6), 1627‚Äì1640.e7. https://doi.org/10.1016/j.celrep.2019.01.041

Jiebiao Wang, Bernie Devlin, Kathryn Roeder, Using multiple measurements of tissue to estimate subject- and cell-type-specific gene expression, Bioinformatics, Volume 36, Issue 3, 1 February 2020, Pages 782‚Äì788, https://doi.org/10.1093/bioinformatics/btz619
